todo_app/view_model.py

```diff
@@ -0,0 +1,56 @@
+from datetime import datetime, timedelta, time
+
+# Decorator for get_cards function to slice result by time.
+def split_result_by_updated_time(func):
+    def wrapper_split_result_by_updated_time(self, list_category, \
+        from_threshold = None, to_threshold = None):
+        intermediate_list = func(self,list_category)
+        final_list = []
+        if from_threshold == None and to_threshold == None:
+            final_list = intermediate_list 
+            return final_list
+        elif from_threshold != None and to_threshold == None:
+            for x in intermediate_list:
+                if x.updated_time >= from_threshold:
+                    final_list.append(x)
+            return final_list
+        elif from_threshold == None and to_threshold != None:
+            for x in intermediate_list:
+                if x.updated_time < to_threshold:
+                    final_list.append(x)
+            return final_list
+        else:
+            for x in intermediate_list:
+                if from_threshold <= x.updated_time < to_threshold: 
+                    final_list.append(x)
+            return final_list
+    return wrapper_split_result_by_updated_time
+
+class ViewModel:
+    def __init__(self, cards: list):
+        self._cards = cards
+
+    @split_result_by_updated_time
+    def get_cards(self, list_category = "All"):
+        result_cards = []
+        if list_category == "All":
+            return self._cards
+        else:
+            for x in self._cards:
+                if x.status == list_category:
+                    result_cards.append(x)
+            return result_cards
+
+    def return_all_done_cards(self, limit_of_cards):
```

Comment:
Ideally, we'd separate out some of this complexity, by having all of the return values as separate properties, rather than combining them all in one list

e.g. one property for `all_done_cards`, another for `show_all_done_cards` etc.

Reply:
This one has me stumped. I read about descriptors and properties but I am unable to apply properties in this context. This is going to be one long-winded answer. 

 **TL;DR:**  What benefit do I get from accessing these return values as properties, that I do not get when I access these as return value of their individual functions (e.g., function show_all_done_cards() as in Step 1)

**Long Version**
- Specifically in the context of this file - I thought properties are only useful if the class attribute needs to have additional logic implemented before it is set (or read, etc.), or if an attribute isn't known at the time when object is instantiated. 
- I tested a few cases as per below -  

> **Step 1:** Let us just break down these return values by putting them in a function of their own. I simplified the file, and looked at just one value, i.e., `show_all_done_cards`
```
class ViewModel:
    def __init__(self, cards: list):
        self._cards = cards

    def get_cards(self):
        return self._cards
    
    def show_all_done_cards(self):
            all_done_cards = self.get_cards()
            if len(all_done_cards) > 5:
                show_all_done_cards = False
            else:
                show_all_done_cards = True
            return show_all_done_cards

viewmodel_obj = ViewModel(["Card 1", "Card 2", "Card 3"])
print("OPTION 1: ", viewmodel_obj.show_all_done_cards())
```
> **Step 2:** I then changed the function to property. As a result, nothing really changes except for the fact that I can now access `show_all_done_cards` as an attribute instead of a function in the print statement. No other difference that I can see ..  
```
    @property
    def show_all_done_cards(self):
            all_done_cards = self.get_cards()
            if len(all_done_cards) > 5:
                show_all_done_cards = False
            else:
                show_all_done_cards = True
            return show_all_done_cards

viewmodel_obj = ViewModel(["Card 1", "Card 2", "Card 3"])
print("OPTION 2: ",viewmodel_obj.show_all_done_cards)
```
> **Step 3:** Desperate to see the benefit of previous step, I added `show_all_done_cards` as attribute. I thought this may be more OOPS-like design and the value can be stored as attribute of the object. 

I then realized that below code does not work because I cannot **set** the value through the **getter** method of property object.

```
class ViewModel:
    def __init__(self, cards: list):
        self._cards = cards
        self._show_all_done_cards = None

    @property
    def show_all_done_cards(self):
            all_done_cards = self.get_cards()
            if len(all_done_cards) > 5:
                show_all_done_cards = False
            else:
                show_all_done_cards = True
            self._show_all_done_cards = show_all_done_cards # Fails here because it can't set the value 
            return self._show_all_done_cards
```

> **Step 4:** Next, I tried to do it the standard way. Write a property, and write a setter as well.  However, I still have questions, and I can't see the benefits of this approach - 

- The below code has a getter and setter. However, the setter has to work before the getter method, otherwise the getter method returns `None`. 
- There is no apparent benefit in storing `show_all_done_cards` as object attribute because the value of this flag may change anytime, hence has to be always calculated through the logic in setter method. Why to store and set this then? 

```
class ViewModel:
    def __init__(self, cards: list):
        self._cards = cards
        self._show_all_done_cards = None

    @property
    def show_all_done_cards(self):
            return self._show_all_done_cards

    @show_all_done_cards.setter
    def show_all_done_cards(self):
            all_done_cards = self.get_cards()
            if len(all_done_cards) > 5:
                show_all_done_cards = False
            else:
                show_all_done_cards = True
            self._show_all_done_cards = show_all_done_cards

viewmodel_obj = ViewModel(["Card 1", "Card 2", "Card 3"])
print("OPTION 4: ",viewmodel_obj.show_all_done_cards)  #This will print None
print(viewmodel_obj.__dict__) # verified object attributes
```